src/zql/context/replicache-context.test.ts

```diff
@@ -0,0 +1,150 @@
+    {id: '1', str: 'a'},
+    {id: '2', str: 'a'},
+    {id: '3', str: 'a'},
+  ]);
+
+  await r.mutate.deleteE1('1');
+
+  expect([...(source as SetSource<E1>).value]).toEqual([
+    {id: '2', str: 'a'},
+    {id: '3', str: 'a'},
+  ]);
+
+  await r.mutate.updateE1({id: '3', str: 'z'});
+
+  expect([...(source as SetSource<E1>).value]).toEqual([
+    {id: '2', str: 'a'},
+    {id: '3', str: 'z'},
+  ]);
+});
+
+test('getSource - with ordering', async () => {
+  const r = newRep();
+  const context = makeReplicacheContext(r);
+  const source = context.getSource('e1', [['str', 'id'], 'asc']);
+
+  await r.mutate.initE1({id: '1', str: 'c'});
+  await r.mutate.initE1({id: '3', str: 'z'});
+  await r.mutate.initE1({id: '2', str: 'a'});
+
+  expect([...(source as SetSource<E1>).value]).toEqual([
+    {id: '2', str: 'a'},
+    {id: '1', str: 'c'},
+    {id: '3', str: 'z'},
+  ]);
+
+  const sourceDesc = context.getSource('e1', [['str', 'id'], 'desc']);
+  // asc/desc don't matter. For desc we just iterate the asc collection backwards.
+  expect(source).toBe(sourceDesc);
+
+  await r.mutate.deleteE1('1');
+  await r.mutate.deleteE1('3');
+
+  expect([...(source as SetSource<E1>).value]).toEqual([{id: '2', str: 'a'}]);
+
+  await r.mutate.updateE1({id: '2', str: 'z'});
+
+  expect([...(source as SetSource<E1>).value]).toEqual([{id: '2', str: 'z'}]);
+});
+
+test('derived sources are correctly seeded', async () => {
+  const r = newRep();
+  const context = makeReplicacheContext(r);
+
+  await r.mutate.initE1({id: '1', str: 'c'});
+  await r.mutate.initE1({id: '3', str: 'z'});
+  await r.mutate.initE1({id: '2', str: 'a'});
+
+  const s1 = context.getSource('e1', [['str', 'id'], 'asc']);
+  const s2 = context.getSource('e1');
+
+  // go on to the next tick of the event loop
+  // so our read to load the source can complete.
+  // Maybe we need something on the source to expose
+  // whether or not it is done seeding.
+  await new Promise<void>(resolve => {
+    setTimeout(resolve, 0);
+  });
+
+  expect([...(s1 as SetSource<E1>).value]).toEqual([
+    {id: '2', str: 'a'},
+    {id: '1', str: 'c'},
+    {id: '3', str: 'z'},
+  ]);
+  expect([...(s2 as SetSource<E1>).value]).toEqual([
+    {id: '1', str: 'c'},
+    {id: '2', str: 'a'},
+    {id: '3', str: 'z'},
+  ]);
+});
+
+// Here as a sanity check for now.
+// Full e2e integration test suite will be in `zql/index.test.ts`
+test('ZQL query with Replicache', async () => {
+  const r = newRep();
+  const context = makeReplicacheContext(r);
+
+  const q = new EntityQuery<{fields: E1}>(context, 'e1');
+
+  const view = q.select('id').where('str', '>', 'm').prepare().materialize();
+
+  await Promise.all([
+    r.mutate.initE1({id: '1', str: 'c'}),
+    r.mutate.initE1({id: '3', str: 'z'}),
+    r.mutate.initE1({id: '2', str: 'a'}),
+    r.mutate.initE1({id: '4', str: 'x'}),
+    r.mutate.initE1({id: '5', str: 'y'}),
+  ]);
+
+  expect(view.value).toEqual([{id: '5'}, {id: '4'}, {id: '3'}]);
```

Comment:
It looks like we are copying the values here, to create objects with only the selected fields. I'm concerned about doing this for perf reasons. I would actually expect this test to fail, because of the extra `str` field in `view.value`.

Reply:
Possible I'm overthinking this. LMK what you think. FYI @arv